smartcosmos-framework/src/main/java/net/smartcosmos/security/user/SmartCosmosUserAuthenticationConverter.java

```diff
@@ -72,32 +73,49 @@ public void setDefaultAuthorities(String[] defaultAuthorities) {
     }
 
     public Authentication extractAuthentication(Map<String, ?> map) {
-        if (map.containsKey(SmartCosmosUser.TOKEN_FIELD_URN_ACCOUNT)) {
-            final String accountUrn = map.get(SmartCosmosUser.TOKEN_FIELD_URN_ACCOUNT)
-                    .toString();
-            final String userUrn = map.get(SmartCosmosUser.TOKEN_FIELD_URN_USER)
-                    .toString();
-            final String username = map.get(USERNAME).toString();
-            Collection<? extends GrantedAuthority> authorities = getAuthorities(map);
-
-            final SmartCosmosUser principal = new SmartCosmosUser(accountUrn, userUrn,
-                    username, "N/A", authorities);
-
-            return new UsernamePasswordAuthenticationToken(principal, "N/A", authorities);
+
+        if (!map.containsKey(USERNAME)) {
+            return null;
         }
-        if (map.containsKey(USERNAME)) {
-            Object principal = map.get(USERNAME);
-            Collection<? extends GrantedAuthority> authorities = getAuthorities(map);
-            if (userDetailsService != null) {
-                UserDetails user = userDetailsService
-                        .loadUserByUsername((String) map.get(USERNAME));
-                authorities = user.getAuthorities();
-                principal = user;
-
-            }
-            return new UsernamePasswordAuthenticationToken(principal, "N/A", authorities);
+
+        if (map.containsKey(SmartCosmosUser.TOKEN_FIELD_URN_ACCOUNT)
+            && map.containsKey(SmartCosmosUser.TOKEN_FIELD_URN_USER)) {
+            return getAuthenticationFromToken(map);
+        } else {
+            return getAuthenticationFromUserDetailsService(map);
         }
-        return null;
+    }
+
+    private Authentication getAuthenticationFromToken(Map<String, ?> map) {
+
+        final String accountUrn = map.get(SmartCosmosUser.TOKEN_FIELD_URN_ACCOUNT)
```

Comment:
opportunity for NPE here if the value is not in the map.  This makes the assumption the caller will verify the map has the values.  This is not a good assumption.  Make the check again in the method it costs next to nothing.  Then you have to determine what to return in the case where there are not correct values.